capirca/lib/fortigate.py

```diff
@@ -0,0 +1,394 @@
+    """
+    f_proto = FortigatePortMap._PROTO_MAP.get(protocol, None)
+    if f_proto is None:
+      raise FortiGateValueError(
+          '%r protocol is unsupported, supported protocols = %r' % (
+              protocol, FortigatePortMap._PROTO_MAP.keys()))
+
+    if isinstance(f_proto, six.string_types):
+      return f_proto
+    elif port:
+      try:
+        return f_proto[port]
+      except KeyError:
+        raise FortiGatePortDoesNotExistError
+    else:
+      raise FortiGateFindServiceError(
+          'service not found from %r protocol and %r port' % (protocol, port))
+
+class ObjectsContainer(object):
+  """A Container that holds service and network objects."""
+
+  def __init__(self):
+    self._fw_addresses = []
+    self._fw_services = []
+
+    self._FW_DUP_CHECK = set()
+
+  def get_fw_addresses(self):
+    """Returns the collected addresses."""
+    self._fw_addresses.extend([' ', 'end', ' '])
+    return self._fw_addresses
+
+  def get_fw_services(self):
+    """Returns the collected services."""
+    self._fw_services.extend([' ', 'end', ' '])
+    return self._fw_services
+
+  def add_address_to_fw_addresses(self, addr):
+    """Add address to address store."""
+    if addr in self._FW_DUP_CHECK:
+      return
+    self._fw_addresses.extend(['\tedit %s' % addr,
+                               '\t\tset subnet %s' % addr,
+                               '\tnext'])
+    self._FW_DUP_CHECK.add(addr)
+
+  def add_service_to_fw_services(self, protocol, service):
+    """Add service to services store."""
+    if service in self._FW_DUP_CHECK:
+      return
+
+    self._fw_services.extend(
+        ['\tedit %s' % service,
+         '\t\tset protocol TCP/UDP',
+         '\t\tset %s-portrange %s' % (protocol.lower(), service),
+         '\tnext'])
+
+    self._FW_DUP_CHECK.add(service)
+
+
+class Term(aclgenerator.Term):
+  """Single Firewall Policy."""
+  ALLOWED_PROTO_STRINGS = ['gre', 'icmp', 'ip', 'tcp', 'udp']
+  COMMENT_MAX_WIDTH = 70
+
+  CURRENT_ID = 0
+
+  def __init__(self, term, object_container):
+    super(Term, self).__init__(term)
+    self._term = term
+    self._obj_container = object_container
+
+    self.id_ = type(self).CURRENT_ID
+    type(self).CURRENT_ID += 1
+
+  @staticmethod
+  def _get_addresses_name(addresses):
+    """Returns the addresses or 'all' if no addresses specified."""
+    v4_addresses = [x.with_prefixlen for x in addresses if
+                    not isinstance(x, nacaddr.IPv6)]
+    addresses = ' '.join(v4_addresses)
+    return addresses or 'all'
+
+  @staticmethod
+  def clean_ports(src_ports, dest_ports):
+    """Returns a set() of src and dest ports."""
+    all_ports = []
+    if src_ports:
+      all_ports += src_ports
+    if dest_ports:
+      all_ports += dest_ports
+    return set(all_ports)
+
+  def _get_services_string(self, protocols, ports):
+    """Get the service name, if not exist create it.
+
+    Args:
+      protocol: list of protocols
```

Comment:
args here do not match the args above, need to add or move the "s" from above or below respectively